actionpack/test/dispatch/request_test.rb

```diff
@@ -116,9 +121,15 @@ class RequestIP < BaseRequestTest
   end
 
   test "remote ip spoof detection" do
-    request = stub_request "HTTP_X_FORWARDED_FOR" => "1.1.1.1",
-                           "HTTP_CLIENT_IP"       => "2.2.2.2"
+    # `stub_request` uses `Rack::Lint` and it has different behavior in Rack 2 and Rack 3.
+    # - Rack 3: an error is raised if it's not a hash
+    # (https://github.com/rack/rack/blob/3-0-stable/lib/rack/lint.rb#L101).
+    # - Rack 2: asserts with `env.inspect` as the assertion message. The type check is only done after evaluating
+    # this `env.inspect` (https://github.com/rack/rack/blob/2-2-stable/lib/rack/lint.rb#L81).
+    # Since ActionDispatch::RemoteIp#to_s calls `calculate_ip` the error is raised immediatly.
     e = assert_raise(ActionDispatch::RemoteIp::IpSpoofAttackError) {
+      request = stub_request "HTTP_X_FORWARDED_FOR" => "1.1.1.1",
+                             "HTTP_CLIENT_IP" => "2.2.2.2"
```

Comment:
Moving the `stub_request` is the easiest thing to do for now. Ideally, `Rack::Lint` in v2 would behave closer to `Rack::Lint` in v3. 

Not sure how to proceed here. I've just left this lengthy comment for now.